### PR TITLE
Stock Agent, Performance improvements, New Retriever

### PR DESCRIPTION
**Pull request by givenscj**
# {PR title}

## The issue or feature being addressed

Fixes #{bug number} (in this specific format)

## Details on the issue fix or feature implementation

{Detail}

## Confirm the following

- [ ]  I started this PR by branching from the head of the default branch
- [ ]  I have targeted the PR to merge into the default branch
- [ ]  I have included unit tests for the issue/feature
- [ ]  I have included inline docs for my changes, where applicable
- [ ]  I have successfully run a local build
- [ ]  I have provided the required update scripts, where applicable
- [ ]  I have updated relevant docs, where applicable

> [!NOTE]
> Instead of adding `X`'s inside the checkboxes you wish to check above, first submit the PR, then check the boxes in the rendered description.

### CODE REVIEW

File: src/python/PythonSDK/foundationallm/models/language_models/language_model.py
Comment (codingbandit) [inline src/python/PythonSDK/foundationallm/models/language_models/language_model.py:1]: Again, this looks like it doesn't have the changes made yesterday.

File: src/python/PythonSDK/foundationallm/models/orchestration/completion_request.py
Comment (codingbandit) [inline src/python/PythonSDK/foundationallm/models/orchestration/completion_request.py:1]: are we using a session id anywhere?

**Review by codingbandit — CHANGES_REQUESTED** [review]

File: .vscode/launch.json
[the remark this answers is not in the document]
Comment (givenscj) [reply .vscode/launch.json:1]: We keep going back and forth.  I work off c:/ so I can comment mine or yours, doesn't matter to me.

File: .github/workflows/release-docker-image.yml
[the remark this answers is not in the document]
Comment (givenscj) [reply .github/workflows/release-docker-image.yml:1]: This was removed.  It came in from previous commits he had in the branch I started with.

File: deploy/arm/azureAcaDeploy.json
[the remark this answers is not in the document]
Comment (givenscj) [reply deploy/arm/azureAcaDeploy.json:1]: Same, reverted to one existing.

File: src/python/LangChainAPI/app/routers/orchestration.py
[the remark this answers is not in the document]
Comment (givenscj) [reply src/python/LangChainAPI/app/routers/orchestration.py:1]: it is temporary until we can figure out why its doing that.

File: src/python/PythonSDK/foundationallm/models/language_models/language_model.py
Comment (givenscj) [reply src/python/PythonSDK/foundationallm/models/language_models/language_model.py:1]: added. conflict resolved.  my config base property is in there for my agent code.

File: .vscode/launch.json
[the remark this answers is not in the document]
Comment (codingbandit) [reply .vscode/launch.json:1]: That D: path isn't mine either ;p

[the remark this answers is not in the document]
Comment (kylebunting) [reply .vscode/launch.json:1]: Should this file even be part of what gets added to the repo? Seems like it is a candidate for being added to the .gitignore file if we are going to see constant conflicts based on how folks have their environment set up.

File: src/dotnet/AgentFactory/Agents/DefaultAgent.cs [hunk 1]
```diff
@@ -71,7 +71,33 @@ public override async Task Configure(string userPrompt, string sessionId)
                             DataDescription = dataSource.DataDescription
                         });
                         break;
+                    case "stock":
+                    //TODO
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Agents/DefaultAgent.cs:75]: Is there more to do here, or does this need to be deleted?

File: src/dotnet/AgentFactory/Agents/DefaultAgent.cs [hunk 2]
```diff
@@ -71,7 +71,33 @@ public override async Task Configure(string userPrompt, string sessionId)
...
+                        {
+                            Name = dataSource.Name,
+                            Type = _agentMetadata.Type,
+                            Description = dataSource.Description,
+                            DataDescription = dataSource.DataDescription,
+                            Configuration = new StockConfiguration
+                            {
+                                IndexName = dataSource.IndexName,
+                                ConfigValueBaseName = dataSource.ConfigValueBaseName,
+                                OpenAIEndpoint = dataSource.OpenAIEndpoint,
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Agents/DefaultAgent.cs:86]: The OpenAI endpoint and key should really be handled via the LanguageModel object that already has this info, and it should be removed from here.

File: src/dotnet/AgentFactory/Agents/DefaultAgent.cs [hunk 3]
```diff
@@ -71,7 +71,33 @@ public override async Task Configure(string userPrompt, string sessionId)
...
+                                OpenAIKey = dataSource.OpenAIKey,
+                                SearchEndpoint = dataSource.SearchEndpoint,
+                                EmbeddingModel = dataSource.EmbeddingModel,
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Agents/DefaultAgent.cs:89]: Embedding model info is passed in via the EmbeddingModel object, so that should be used and this should be removed from here.

File: src/dotnet/AgentFactory/Agents/DefaultAgent.cs [hunk 4]
```diff
@@ -71,7 +71,33 @@ public override async Task Configure(string userPrompt, string sessionId)
...
+                                SearchKey = dataSource.SearchKey,
+                                ConnectionStringSecretName = dataSource.ConnectionStringSecret,
+                                ContainerName = dataSource.Container,
+                                RetrieverMode = dataSource.RetrieverMode,
+                                LoadMode = dataSource.LoadMode,
+                                Company = dataSource.Company,
+                                Sources = dataSource.Sources
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Agents/DefaultAgent.cs:97]: Shouldn't this be handled via the DataSources object? I'm confused as to what this is doing here.

File: deploy/scripts/BuildPush.ps1
Comment (kylebunting) [inline deploy/scripts/BuildPush.ps1:1]: I'm not sure why these files are in here, so cannot really comment or review them.

File: src/dotnet/AgentFactory/Models/Messages/DataSourceHubResponse.cs [hunk 5]
```diff
@@ -139,5 +139,39 @@ public record DataSourceMetadata
         /// </summary>
         [JsonProperty("text_field_name")]
         public string? TextFieldName { get; set; }
+
+        [JsonProperty("open_ai_endpoint")]
+        public string? OpenAIEndpoint { get; set; }
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Models/Messages/DataSourceHubResponse.cs:144]: Same comments as above about the OpenAI and Embeddings models being handled elsewhere.

File: src/dotnet/AgentFactory/Models/Orchestration/DataSourceConfigurations/StockConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace FoundationaLLM.AgentFactory.Core.Models.Orchestration.DataSourceConfigurations
+{
+    /// <summary>
+    /// Blob storage configuration settings.
+    /// </summary>
+    public class StockConfiguration
+    {
+        [JsonProperty("open_ai_endpoint")]
+        public string? OpenAIEndpoint { get; set; }
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Models/Orchestration/DataSourceConfigurations/StockConfiguration.cs:16]: Same comment as above.

File: src/dotnet/AgentFactory/Models/Orchestration/Metadata/StockDataSource.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace FoundationaLLM.AgentFactory.Core.Models.Orchestration.Metadata
+{
+    /// <summary>
+    /// Blob storage data source metadata model.
+    /// </summary>
+    public class StockDataSource : DataSourceBase
+    {
+        /// <summary>
+        /// Blob storage configuration settings.
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Models/Orchestration/Metadata/StockDataSource.cs:12]: You should fix your XML inline doc to reflect the Stock, not blob storage, config.

File: src/dotnet/AgentFactory/Services/AgentFactoryService.cs [hunk 8]
```diff
@@ -82,6 +82,17 @@ public async Task<CompletionResponse> GetCompletion(CompletionRequest completion
...
+            if (completionRequest.UserPrompt == "--warmup--")
+                return new CompletionResponse
+                {
+                    Completion = "Warmed up.",
+                    UserPrompt = completionRequest.UserPrompt ?? string.Empty,
+                    PromptTokens = 0,
+                    CompletionTokens = 0,
+                    UserPromptEmbedding = new float[] { 0 }
+                };
+
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Services/AgentFactoryService.cs:94]: Remove the extra lines/whitespace?

File: src/dotnet/AgentFactory/Services/AgentFactoryService.cs [hunk 9]
```diff
@@ -82,6 +82,17 @@ public async Task<CompletionResponse> GetCompletion(CompletionRequest completion
                 _promptHubAPIService,
                 _dataSourceHubAPIService);
 
+            if (completionRequest.UserPrompt == "--warmup--")
```
Comment (kylebunting) [inline src/dotnet/AgentFactory/Services/AgentFactoryService.cs:85]: What purpose does this serve, exactly?

File: src/dotnet/AgentFactory/Services/AgentHubAPIService.cs
Comment (kylebunting) [inline src/dotnet/AgentFactory/Services/AgentHubAPIService.cs:1]: Are we introducing a new risk by adding this caching functionality right now?

File: src/dotnet/AgentFactory/Services/DataSourceHubAPIService.cs
Comment (kylebunting) [inline src/dotnet/AgentFactory/Services/DataSourceHubAPIService.cs:1]: Same comment as above about risk with the caching functionality.

File: src/dotnet/AgentFactory/Services/PromptHubAPIService.cs
Comment (kylebunting) [inline src/dotnet/AgentFactory/Services/PromptHubAPIService.cs:1]: Same as comments above about caching

File: src/python/IntegrationSDK/foundationallm/integration/config/configuration.py [hunk 10]
```diff
@@ -26,15 +26,15 @@ def get_value(self, key: str) -> str:
...
+
         Returns
         -------
         The configuration value
 
         Raises an exception if the configuration value is not found.
         """
         app_config_uri = os.environ['foundationallm-app-configuration-uri']
-        credential = DefaultAzureCredential()
+        credential = DefaultAzureCredential( exclude_environment_credential=True)
```
Comment (kylebunting) [inline src/python/IntegrationSDK/foundationallm/integration/config/configuration.py:37]: Why do we need to add this now?

File: src/python/LangChainAPI/app/routers/orchestration.py [hunk 11]
```diff
@@ -35,8 +38,23 @@ async def get_completion(completion_request: CompletionRequest,
         Object containing the completion response and token usage details.
     """
     try:
+        #get the orginal request content
```
Comment (kylebunting) [inline src/python/LangChainAPI/app/routers/orchestration.py:41]: This whole block needs to not be here. The Python APIs should be lightweight and not have any logic about a specific agent type in them.

File: src/python/PythonSDK/foundationallm/config/configuration.py [hunk 12]
```diff
@@ -12,11 +12,14 @@ class Configuration():
     def __init__(self):
         """Init"""
         try:
-            app_config_uri = os.environ['foundationallm-app-configuration-uri']
+            if (os.environ['ENV'] == "development"):
+                app_config_uri = os.environ['foundationallm-app-configuration-uri-dev']
+            else:
+                app_config_uri = os.environ['foundationallm-app-configuration-uri']
         except Exception as e:
-            raise e
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/config/configuration.py:17]: Shouldn't we still be raising an exception here? What happens in the setting is missing from the local environment variables?

File: src/python/PythonSDK/foundationallm/config/configuration.py [hunk 13]
```diff
@@ -12,11 +12,14 @@ class Configuration():
...
+            app_config_uri = os.environ['foundationallm-app-configuration-uri']
 
-        credential = DefaultAzureCredential()
+        credential = DefaultAzureCredential( exclude_environment_credential=True)
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/config/configuration.py:22]: Why is this needed now?

File: src/python/PythonSDK/foundationallm/hubs/agent/agent_hub.py [hunk 14]
```diff
@@ -4,9 +4,14 @@
 
 class AgentHub(HubBase):
     """The AgentHub is responsible for resolving agents."""
-    def __init__(self):
+    def __init__(self, config=None):
+
         # initialize config
-        self.config = Configuration()
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/hubs/agent/agent_hub.py:9]: You should be able to shorten this to be:

`self.config = config or Configuration()`

File: src/python/PythonSDK/foundationallm/hubs/data_source/data_source_hub.py [hunk 15]
```diff
@@ -5,9 +5,14 @@
 class DataSourceHub(HubBase):
     """The DataSourceHub is responsible for resolving data sources."""
 
-    def __init__(self):
+    def __init__(self, config=None):
+
         # initialize config
-        self.config = Configuration()
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/hubs/data_source/data_source_hub.py:10]: Same comment as above. Make this:

`self.config = config or Configuration()`

File: src/python/PythonSDK/foundationallm/hubs/data_source/data_sources/stock/stock_data_source_metadata.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
+from typing import List, Optional
+from foundationallm.hubs.data_source import DataSourceMetadata
+
+class StockDataSourceMetadata(DataSourceMetadata):
+    """Configuration settings for a stock agent"""
+    open_ai_endpoint: str
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/hubs/data_source/data_sources/stock/stock_data_source_metadata.py:6]: Same comments as above for the OpenAI and Embeddings settings coming in with the Language and Embedding models. These values should be pulled from there so we aren't configuring the setting in multiple places.

File: src/python/PythonSDK/foundationallm/hubs/prompt/prompt_hub.py [hunk 17]
```diff
@@ -5,7 +5,12 @@
 class PromptHub(HubBase):
     """The PromptHub class is responsible for resolving prompts."""
 
-    def __init__(self):
-         # initialize config
-        self.config = Configuration()
+    def __init__(self, config=None):
+
+        # initialize config
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/hubs/prompt/prompt_hub.py:10]: Make this:

`self.config = config or Configuration()` to replace the if statement.

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py [hunk 18]
```diff
@@ -0,0 +1,534 @@
...
+        )
+
+        self.blob_service_client = self.storage_manager.blob_service_client
+
+        self.search_endpoint = config.get_value(self.data_source.search_endpoint)
+        self.search_key = config.get_value(self.data_source.search_key)
+
+        self.vector_store_address = config.get_value(self.data_source.search_endpoint)
+        self.vector_store_password = config.get_value(self.data_source.search_key)
+
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:86]: The settings for the azure endpoint and key should be pulled from the LanguageModel object.

The embeddings model, should come from the EmbeddingModel.

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py [hunk 19]
```diff
@@ -0,0 +1,534 @@
...
+        for source in temp_sources:
+            self.filters.append(f"search.ismatch('{source}', 'metadata', 'simple', 'all')")
+
+        if ( retriever_mode == "azure" ):
+            local_path = f"{company}-financials"
+            retriever = self.get_azure_retiever(local_path, embedding_field_name="content_vector", text_field_name="content", top_n=20)
+
+        if ( retriever == "chroma" ):
+            local_path = f"c:/temp/{company}"
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:119]: Is this for local testing? When are we going to have something referencing c:?

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py [hunk 20]
```diff
@@ -0,0 +1,534 @@
...
+            return CompletionResponse(
+                completion = self.llm_chain.invoke(self.question, return_only_outputs=True)['answer'],
+                user_prompt = self.question,
+                completion_tokens = cb.completion_tokens,
+                prompt_tokens = cb.prompt_tokens,
+                total_tokens = cb.total_tokens,
+                total_cost = cb.total_cost
+            )
+
+    def who_is_question_for(self, question):
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:198]: We cannot have this in here! This is essentially hardcoding responders, which we don't want.

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py [hunk 21]
```diff
@@ -0,0 +1,534 @@
...
+        pdf_retriever = self.load_pdfs(local_path, retriever_mode=retriever_mode, load_mode=load_mode)
+
+        docs_10k_retriever = self.load_10k(local_path, retriever_mode=retriever_mode, load_mode=load_mode)
+
+        docs_10q_retriever = self.load_10q(local_path, retriever_mode=retriever_mode, load_mode=load_mode)
+
+        docs_14a_retriever = self.load_14a(local_path, retriever_mode=retriever_mode, load_mode=load_mode)
+
+        transcripts_retriever = self.load_transcripts(local_path, retriever_mode=retriever_mode, load_mode=load_mode)
+
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:327]: These all reference a specific company and doc location. We don't want this in here.

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py [hunk 22]
```diff
@@ -0,0 +1,534 @@
...
+        print(f'Getting retriever {retriever_mode}')
+
+        if ( retriever_mode == "azure" ):
+
+            credential = AzureKeyCredential(self.vector_store_password)
+
+            retriever = SearchServiceRetriever(
+                    endpoint=self.vector_store_address,
+                    index_name="cjg-vector-index",
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:371]: We probably should not be hardcoding an index name.

File: src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/agents/stock_agent.py:1]: This file has a lot of hardcoded values to things. We need to get that fixed before it can be merged.

File: src/python/PythonSDK/foundationallm/langchain/data_sources/stock/stock_configuration.py [hunk 23]
```diff
@@ -0,0 +1,24 @@
+from typing import List
+from foundationallm.langchain.data_sources import DataSourceConfiguration
+
+class StockConfiguration(DataSourceConfiguration):
+    """Configuration settings for a stock agent"""
+    open_ai_endpoint: str
```
Comment (kylebunting) [inline src/python/PythonSDK/foundationallm/langchain/data_sources/stock/stock_configuration.py:6]: Same comments as above about these values and the Language Model

File: tests/python/PythonSDK.Tests/config/configuration_tests.py [hunk 24]
```diff
@@ -25,10 +25,9 @@ def test_configuration_includes_feature_flags(self, test_config):
         actual = test_config.get_feature_flag("FoundationaLLM-AllowAgentHint")       
         assert actual == True
         
-
     def test_config_setting_change(self):        
         app_config_uri = os.environ['foundationallm-app-configuration-uri']
-        client = AzureAppConfigurationClient(app_config_uri, DefaultAzureCredential())
+        client = AzureAppConfigurationClient(app_config_uri, DefaultAzureCredential( exclude_environment_credential=True))
```
Comment (kylebunting) [inline tests/python/PythonSDK.Tests/config/configuration_tests.py:30]: Why do we need this now?

**Review by kylebunting — CHANGES_REQUESTED** [review]
